tests/RobotFramework/tests/cumulocity/registration/device_deregistration.robot

```diff
@@ -0,0 +1,103 @@
+*** Settings ***
+Resource    ../../../resources/common.resource
+Library    Cumulocity
+Library    ThinEdgeIO
+
+Test Tags    theme:c8y    theme:registration
+Suite Setup    Custom Setup
+Test Setup    Test Setup
+Test Teardown    Get Logs    ${DEVICE_SN}
+
+*** Test Cases ***
+
+Main device registration
+    ${mo}=    Device Should Exist              ${DEVICE_SN}
+    ${mo}=    Cumulocity.Device Should Have Fragment Values    name\=${DEVICE_SN}
+    Should Be Equal    ${mo["owner"]}    device_${DEVICE_SN}
+    Should Be Equal    ${mo["name"]}    ${DEVICE_SN}
+
+
+Child device registration
+    Execute Command    tedge mqtt pub --retain 'te/device/${CHILD_SN}//' '{"@type":"child-device","@id":"${CHILD_SN}"}'
+
+    # Check registration
+    ${child_mo}=    Device Should Exist        ${CHILD_SN}
+    ${child_mo}=    Cumulocity.Device Should Have Fragment Values    name\=${CHILD_SN}
+    Should Be Equal    ${child_mo["owner"]}    device_${DEVICE_SN}    # The parent is the owner of the child
+    Should Be Equal    ${child_mo["name"]}     ${CHILD_SN}
+
+    # Check child device relationship
+    Cumulocity.Set Device    ${DEVICE_SN}
+    Cumulocity.Should Be A Child Device Of Device    ${CHILD_SN}
+
+    #Deregister Child device
+    Run Keyword And Ignore Error    Execute Command    mosquitto_sub --remove-retained -W 3 -t 'te/device/${CHILD_SN}//'
+    
```

Comment:
One way to validate the de-registration is to deregister the device as done here, restart the mapper and then try re-registering it again with the same registration payload and validate if the corresponding 101 SmartREST message is published by the mapper. If you wanna do it completely via C8y, then you can delete the child device from C8y as well, after the deregistration step and then validate if the child device is recreated after the re-registration.

Once we implement the proper deregistration handling in the mapper (which removes the entries from the in-memory entity store on receipt of the clear messages), then the mapper restart won't be needed.

Reply:
Yeah, the failure is a result of using timeouts (the `-W` option) to end the command. But, `mosquitto_sub` exits with exit code 27 when it is a timeout. So, you can selectively ignore exit code 27, instead of ignoring everything to be more confident that the command did work.

Reply:
So this  could be refactored to:

```
Execute Command    mosquitto_sub --remove-retained -W 3 -t 'te/device/${CHILD_SN}//'    exp_exit_code=27
```